### PR TITLE
Try adding a 'spotlight mode' type effect when template part or child is selected.

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
## Description

Re-exploring the idea of adding a 'spotlight mode' effect on template parts.
Here, we add a new class to pass down from the block-list if a template part or any of its children are selected.  And some style rules similar to the current focus-mode.

This works as a default behavior for template part/child selection and the styles are overridden if 'spotlight mode' is enabled.

## Screenshots 
spotlight-mode-template-part-2

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by Addison-Stavlo** [comment]
Any thoughts on how this feels @shaunandrews @jameskoster ?  It will be overridden if 'spotlight mode' becomes enabled, but I don't think thats necessarily a bad thing.

**Comment by dubielzyk** [comment]
I quite like it. We're not communicating why we're entering spotlight mode to the user, but perhaps that's fine. Would this also trigger if you have nested template parts?

**Comment by jameskoster** [comment]
I think spotlight is the way to go, but the finer details of what happens when inner blocks are selected, or when multiple instances of a single template part exist on the canvas, needs some refinement.

In #25452 I worked on a couple of crude prototypes that we can perhaps discuss here.

In this [latest iteration](https://cdpn.io/jameskoster/debug/XWdGWBr):

* The selected template part is outlined and spotlighted. The outline helps when spotlight mode is enabled.
* Any other instances of that template part are outlined but excluded from the spotlight.
* When an inner block is selected the entire parent template part remains spotlighted and equivalent inner blocks in other instances of the same template part are also spotlighted and outlined.

**Comment by Addison-Stavlo** [comment]
> Would this also trigger if you have nested template parts?

@dubielzyk - Good question, it is a little odd right now the nested template parts are spotlighted together regardless of whether the parent or child is selected.  I will try to make this better so that only the child is spotlit when it is the one selected.

> when multiple instances of a single template part exist on the canvas, needs some refinement.

@jameskoster  - I think it is fair to follow up and reiterate regarding how to handle duplicates of the same template part that aren't selected.  Currently, it isn't possible to have 2 of the same template part on one page and that bug is in the process of being worked out technically.  But should that prevent us from adding the ability to spotlight for the selected template part in the meantime?  If we agree that the opacity changes are the direction forward, then a small scoped change like this would be a good first step.

> * The selected template part is outlined and spotlighted. The outline helps when spotlight mode is enabled.
> * Any other instances of that template part are outlined but excluded from the spotlight.
> * When an inner block is selected the entire parent template part remains spotlighted and equivalent inner blocks in other instances of the same template part are also spotlighted and outlined.

The first outline already exists as a subtle gray border and I agree it does work well with the template part spotlight mode.  The second would be a simple change in the blocks editor files, and the third would require some changes in the block-list/wrappers but should be doable as well if we determine that is the way to go.  Both of these I could see as potential follow ups after we are able to successfully render 2 of the same template parts on one page.

**Comment by jameskoster** [comment]
> Both of these I could see as potential follow ups after we are able to successfully render 2 of the same template parts on one page.

Seems fair to me :)

**Comment by shaunandrews** [comment]
This feels nice. If we can make the logic work for nested template parts, I'd love to see this merged.

**Comment by Addison-Stavlo** [comment]
> This feels nice. If we can make the logic work for nested template parts, I'd love to see this merged.

Nice!  848b797 + dcc2d5f should have this working for the nested cases now. 😁

**Comment by noahtallen** [comment]
Nice, the nested case works well. I also appreciate how "spotlight mode" takes precedent over this, so it still works the same as before with template part blocks.

File: packages/block-editor/src/components/block-list/index.js
Comment (noahtallen) [inline packages/block-editor/src/components/block-list/index.js:141]: Will this work if `activeTemplatePartId` has ID 0?

File: packages/block-editor/src/components/block-list/index.js
Comment (noahtallen) [inline packages/block-editor/src/components/block-list/index.js:55]: It feels like this could be expensive to run all the time when template parts don't exist. I wonder if it would be possible to avoid running this code when the template part block has not been registered (That would keep it out of non-fse sites for now)

**Review by shaunandrews — APPROVED** [review]

File: packages/block-editor/src/components/block-list/style.scss
Comment (noahtallen) [inline packages/block-editor/src/components/block-list/style.scss:95]: Can we remove some of these classnames since we can assume that `is-active-template-part` means that it's selected?

**Review by noahtallen — COMMENTED** [review]
Nice! I like how this works.

before merging, i'd love to see if we could disable it for non-FSE sites (so we can test it out without affecting plugin consumers).

Besides that, from a code point of view, I dislike how we are tightly coupling one block type with the general block list. I wonder if there's a better way to handle it -- perhaps adding `is-active-template-part` to the template part block's edit method (or just adding the right CSS properties to the template part's edit div)

I can see how `template-part-highlighting` needs to be more general though. I'm not thinking of any alternatives to that approach off the top of my head though.

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:141]: The ID generally a string, so `"0"` would be truthy.  If it were a numeric value, it would not work.  However this seems a bit different from the standard convention for blockIds, but we could account for that if it seems necessary?

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:55]: > It feels like this could be expensive to run all the time when template parts don't exist.

I didn't think this would be very expensive?  And its up at the block-list level so its not running for every block.

> I wonder if it would be possible to avoid running this code when the template part block has not been registered

Yeah, I think we could gate it behind the experiment being enabled.

File: packages/block-editor/src/components/block-list/index.js
Comment (noahtallen) [reply packages/block-editor/src/components/block-list/index.js:141]: Gotcha! I didn't realize it was a string here :)

File: packages/block-editor/src/components/block-list/index.js
Comment (noahtallen) [reply packages/block-editor/src/components/block-list/index.js:55]: > I didn't think this would be very expensive

I guess it'd just be expensive for deeply nested blocks, which would have to parse all the way up the block tree to get a result. But I think it's memoized so that helps. I also didn't realize it wasn't running for every block which is nice :) So it's probably me just being overly cautious. Definitely fine to try out in the site editor 👍

**Comment by Addison-Stavlo** [comment]
> I dislike how we are tightly coupling one block type with the general block list. I wonder if there's a better way to handle it 

I understand that concern.  However, we do need to access this area to handle adding the `template-part-highlighting` class to blocks in the editor.  So using this as the place to add `is-active-template-part` consolidates the code for this functionality to one area.  Similar to how 'spotlight mode' adds `is-focus-mode` and `is-focused` here, we are adding the equivalent versions of that for template parts.

While it is 'just one block type', it could be extended to more block type if desired in the future (post content, etc.).  So maybe rebranding these to more generic terms than 'template parts' would make sense? 🤷‍♀️ 

> -- perhaps adding `is-active-template-part` to the template part block's edit method 

We could do that, although it would separate the code used to achieve this functionality into 2 separate packages which to me seems much less ideal.

> (or just adding the right CSS properties to the template part's edit div)

Im not sure I follow what these right CSS properties would be?  Opacity is a bit of a complex case in that it bleeds down, so just setting `opacity: 1` on the template part won't do much if it has a parent already set down to 0.5.  The styles to achieve this opacity reduction without multiplying the effect on nested blocks is very well bundled together at the moment.  If we move all the styles to template part edit, it doesn't make much sense for the `.template-part-highlighting` to be defined in one package and have its styles defined in another.  If we decouple the styles and move them across 2 different packages, that makes this much more difficult to manage in the future.

In the end, using the block-list / wrappers seems necessary for implementing an editor wide content opacity control and while we may be referencing 'one specific block type' in the general list component that seems much more manageable than splitting up the functionality and style rules between multiple packages.

File: packages/block-editor/src/components/block-list/style.scss
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/style.scss:95]: Ah yeah we probably don't need that one anymore 😁

**Comment by noahtallen** [comment]
Thanks for the detailed explanation!

> So maybe rebranding these to more generic terms than 'template parts' would make sense

I would be in favor of this then!

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js]: Gated this to check for the site editor experiment - bea6a37

File: packages/block-editor/src/components/block-list/style.scss
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/style.scss:95]: Looking again, I don't think we can remove any of these selectors and still have it functioning properly.

We could replace the `&.is-selected` with `&.is-active-template-part` if that seems more ideal to you?

**Comment by Addison-Stavlo** [comment]
> > So maybe rebranding these to more generic terms than 'template parts' would make sense

> I would be in favor of this then

Any ideas on names? 😆   `is-active-entity`/`entity-highlighting` ? or instead of 'entity' other ideas like content, context, region... 🤔

**Comment by noahtallen** [comment]
ideas:

- `is-active-entity`/`has-active-entity` (changing the second one) 
- `is-active-block-area`/`has-active-block-area`
- `is-active-controller`/`has-active-controller` (like inner block controller)

File: packages/block-editor/src/store/selectors.js
Comment (Addison-Stavlo) [inline packages/block-editor/src/store/selectors.js:543]: Stealing all this from @noahtallen 's nice addition on #25547 that never went through for other reasons.  Its definitely helpful. 😁

**Review by Addison-Stavlo — COMMENTED** [review]
I updated this to be capable of easily adding other entity blocks in the future if necessary (such as 'core/post-content'), and updated the classnames to reflect this appropriately `is-active-entity`/`has-active-entity`.  (Nitpicking `has-active-entity` makes me think that the active entity would be a child of the element that has this class on it, when it is really just somewhere in the editor.  Although, this is better than other suggestions and I can't think of anything better.)

File: packages/block-editor/src/components/block-list/style.scss
Comment (noahtallen) [reply packages/block-editor/src/components/block-list/style.scss:95]: I personally think that makes sense, since only the active template part would receive opacity: 1, but it's up to you. Since it's functionally important we don't have to change it

File: packages/block-editor/src/components/block-list/index.js
Comment (noahtallen) [inline packages/block-editor/src/components/block-list/index.js:26]: I do like how this feels more extensible without adding much extra programming complexity

**Review by noahtallen — APPROVED** [review]
nice! I think this works well, and it the changes make sense to me.

**Comment by Addison-Stavlo** [comment]
Great!  We could use a core team member to approve since we are making these changes in the block list/wrapper (although still gated behind the FSE experiment being active) as well as updating that selector.  

@youknowriad - if you get some time spare to sanity check this one.

File: packages/block-editor/src/components/block-list/index.js
Comment (youknowriad) [reply packages/block-editor/src/components/block-list/index.js:26]: I don't like this personally, it's static and references blocks that might or might not exist in the available blocks. Can't we rely on the "controlled" inner blocks instead, we have a flag for that I believe.

File: packages/block-editor/src/components/block-list/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/index.js:59]: Why are we restricting to FSE here? I can see a world where this is useful for reusable blocks if they are refactored to work like template parts.

File: packages/block-editor/src/components/block-list/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/index.js:82]: This is a bit too complex for `mapSelectToProps`, should this be moved to a selector?

File: packages/block-editor/src/store/selectors.js
Comment (youknowriad) [reply packages/block-editor/src/store/selectors.js:543]: what is this change about?

**Comment by youknowriad** [comment]
did we try to see how this combines with the global spotlight mode?

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:82]: I was just thinking this as well last night, that makes sense to me.

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:59]: Mostly out of caution for the time being as the only block currently being tested in this is the template part.  But I agree, this could definitely be useful for non-FSE things.  Its just that the designs that have driven this to life have only thought about template parts for now.

**Comment by Addison-Stavlo** [comment]
> did we try to see how this combines with the global spotlight mode?

Yup!  global spotlight mode overrides this, which seems to make sense and has good behavior.  So if you are working in an active template part, the entire template part is spotlit.  When you turn on global spotlight mode, only the selected block is spotlit as expected.

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:26]: > might or might not exist in the available blocks

I guess that is one reason running the code that uses this is gated behind the FSE experiment being active for now.

> Can't we rely on the "controlled" inner blocks instead, we have a flag for that I believe.

If we only want to test this for template parts currently, and have the ability to test/add it with other entity controller blocks one at a time, evaluating based on controlled inner blocks would be a bit too general?

Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:26]: If we are moving the mapSelect logic to a selector, Im thinking this list of allowed blocks to enable this feature will go there with it.

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:82]: Also noting moving this logic to a selector will allow us to clean up the document-actions header component in edit-site since it could use this information as well.

File: packages/block-editor/src/store/selectors.js
Comment (youknowriad) [reply packages/block-editor/src/store/selectors.js:543]: Let's a unit test to this selector to cover that

File: packages/block-editor/src/components/block-list/index.js
Comment (youknowriad) [reply packages/block-editor/src/components/block-list/index.js:26]: > evaluating based on controlled inner blocks would be a bit too general?

Right now, the only "controlled" blocks are "Post content" or "Template part" and I believe later we'll have "Reusable blocks". All these blocks seem like good candidates for this behavior for me.

Comment (noahtallen) [reply packages/block-editor/src/components/block-list/index.js:26]: Checking if it's an inner block controller does make sense to me. I personally haven't been sure how far we should rely on that for this information, but I do like the simplicity.

The only challenge would be if a 3rd party user creates a block which is a controller, but doesn't expect this UI side effect. I don't think that needs to be a big concern for now, though.

File: packages/block-editor/src/components/block-list/style.scss
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/style.scss]: updated

File: packages/block-editor/src/store/selectors.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/store/selectors.js:543]: updated

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:26]: Regarding detecting by controlled inner blocks - while I do think that would clean things up and that the other controlled inner blocks are good candidates for this behavior, I don't think the design discussions have gotten much of a chance to think about this.

Moreover, detecting by a list of allowed blocks allows us to easily re-use the new selector in other areas that have different needs.  For example, I cleaned up the `packages/edit-site/src/components/header/document-actions/index.js` `useSecondaryText` hook quite a bit by being able to reuse this logic.  In that component, it was argued to only show the title for active template parts and not post content.

So if we wanted to try this spotlight behavior for post content we could just update the selection to be
`getActiveBlockIdByBlockNames( [ 'core/template-part', 'core/post-content' ] )`, yet still have the previously mentioned area just using `getActiveBlockIdByBlockNames( [ 'core/template-part' ] )` (or vice versa).

This format also allows for future usage outside controlled blocks in general.  The new selector could be used for other blocks such as group block, columns block, or whatever else might be useful in future development.

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js]: updated

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:56]: Now it is just there because the only block in the list is of allowed blocks is only registered in FSE, so it stops the selector from running unnecessarily.  We could remove this if desired, or if other blocks are added to this functionality.

**Comment by Addison-Stavlo** [comment]
Updated this to:
* Create a new selector.
    * use this selector to clean up `useSecondaryText` in document-actions.
*  Add tests for the updated and new selectors.
*  Updated some redundant or unnecessary style rules.

could use another look and testing.

File: packages/block-editor/src/store/selectors.js
Comment (noahtallen) [inline packages/block-editor/src/store/selectors.js:1845]: Not sure if this is needed, but I think we could add `valildBlockNames` to the list here just in case.

File: packages/edit-site/src/components/header/document-actions/index.js
Comment (noahtallen) [inline packages/edit-site/src/components/header/document-actions/index.js:34]: nice, this is awesome :D

File: packages/block-editor/src/components/block-list/index.js
Comment (noahtallen) [reply packages/block-editor/src/components/block-list/index.js:56]: I think the selector still has to do the same "expensive" calculations to figure out if a block is not activee

File: packages/block-editor/src/store/selectors.js
Comment (noahtallen) [inline packages/block-editor/src/store/selectors.js:1819]: I'm not a huge fan of this name, but I can't think of anything better either 🙃 Maybe we should mark it as __experimental so that people don't start using it in the meantime?

**Comment by noahtallen** [comment]
it also is still testing great for me locally 👍

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:56]: > I think the selector still has to do the same "expensive" calculations to figure out if a block is not active

Ah gotcha.  It should never change either, so maybe we run it once to get the value on mount or something?  Or... just get rid of it?

File: packages/block-editor/src/store/selectors.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/store/selectors.js]: Agreed, and sure that makes sense to me!  done

File: packages/block-editor/src/store/selectors.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/store/selectors.js]: Added. ✅    -  Yeah, probably not needed... but if someone may find an interesting use for this selector that requires that.

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:56]: Or maybe it would make sense to make an `__experimentalIsFullSiteEditingActive` selector that is memoized? 🤔

Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:56]: > I think the selector still has to do the same "expensive" calculations to figure out if a block is not active

Im not sure I follow.  `getSettings` ? Or I guess the new selector since you are mentioning the block being active? Luckily we aren't calling it on/for every block, just on the block list and only for the selected block. 🤔

Comment (noahtallen) [reply packages/block-editor/src/components/block-list/index.js:56]: > Now it is just there because the only block in the list is of allowed blocks is only registered in FSE, so it stops the selector from running unnecessarily

I actually misunderstood what you were saying! I read this as "the list of allowed blocks stops the selector from running unnecessarily because the allowed blocks are only registered when FSE is active". So you can disregard my comment 😅

File: packages/block-editor/src/components/block-list/index.js
Comment (youknowriad) [reply packages/block-editor/src/components/block-list/index.js:26]: I still disagree about having this list in `block-editor` personally. If this is to be a curated list, I think it needs to be a setting of `BlockEditorProvidder` but to be honest, I don't see why design would not want post content to not use spotlight mode like template part, in some ways, it's just a "content" template part. 

If you prefer we can start with an unstable block editor setting?

Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:26]: > If you prefer we can start with an unstable block editor setting?

I think that would definitely be better than what we are currently doing.  Would that be to define the list [in the store defaults](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/store/defaults.js) with an `__experimental` prefix?

> I don't see why design would not want post content to not use spotlight mode like template part, in some ways, it's just a "content" template part.

I agree, maybe we can update this from a list in settings to checking for controlled inner blocks in the future instead if this is desirable.  

@shaunandrews @kellychoffman @jameskoster @dubielzyk @olaolusoga - Has there been much discussion or thought regarding using this sort of spotlight mode for Post Content (and in the future reusable blocks) as well?  Similarly with what shows up as the secondary/selected item in the document-actions name panel in the top bar?  I get the feeling that there should be consistency between what we show in the top-bar and what we highlight with the 'spotlight' mode.

Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js]: @youknowriad - Updated to pass the list into settings from the site editor's initialization, with a default empty array on the block-editors side.  The selector will also return early if there is nothing in that array so we got rid of the check for the experiment being active as well. (thanks to some help from @noahtallen 🎉 )

**Review by noahtallen — APPROVED** [review]
I like the new changes. Setting that via settings seems to be a lot cleaner, and it helps us avoid running any of the (theoretically expensive?) code outside of the site editor for now.

It also works very well for me locally:



The only thing I'm less sure about is the names of the selectors and state we've added, but since we've marked them as experimental, we have time to improve them if folks have better ideas :)

File: packages/block-editor/src/components/block-list/style.scss
Comment (vindl) [inline packages/block-editor/src/components/block-list/style.scss:89]: We could update this to refer to spotlight entities and mention template parts as one specific example.

File: packages/block-editor/src/components/block-list/index.js
Comment (vindl) [inline packages/block-editor/src/components/block-list/index.js:53]: Would it make sense to make this a default parameter value if it's not passed? Then we could also probably shorten the name of the selector to `__experimentalGetActiveEntityBlockId`, which better aligns it with surrounding variable names that we are using. I guess it depends on how often we anticipate this selector will be used outside of this functionality.

Also I'd probably rename `__experimentalSpotlightBlocks` to `__experimentalSpotlightEntityBlocks` to avoid potential confusion with regular spotlight feature.

**Review by vindl — APPROVED** [review]
I really like the latest round of updates to this, the code looks much more polished now! ✨ Gave it another spin and everything still works as expected on my end. I also made sure that these changes are not affecting the post editor.

I added a few optional naming suggestions.

File: packages/block-editor/src/store/selectors.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/store/selectors.js:523]: 🤔   Im not sure tbh.  The original use is a string of a single blockName and this is still supported.  Since it can either be a string or an array of strings, whether we use `blockName` or `blockNames` it will still read a bit funky somewhere. ( like `name === blockNames` 🤷‍♀️

File: packages/block-editor/src/components/block-list/index.js
Comment (Addison-Stavlo) [reply packages/block-editor/src/components/block-list/index.js:53]: > Would it make sense to make this a default parameter value if it's not passed?
> I guess it depends on how often we anticipate this selector will be used outside of this functionality.

Yeah, I guess that 100% depends on how this selector might be used.  For our purposes, that would make sense!  However, without that change this selector is more general and could be used for other purposes.  Whether it would be, I really don't know. 🤷‍♀️ 

> Also I'd probably rename __experimentalSpotlightBlocks to __experimentalSpotlightEntityBlocks to avoid potential confusion with regular spotlight feature.

👍

**Review by vindl — APPROVED** [review]
LGTM! 🚀

**Pull request merged**
